File: native/src/socket_options.c
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

#include "crt.h"
#include "exports.h"

#include <aws/io/socket.h>

AWS_DOTNET_API struct aws_socket_options *aws_dotnet_socket_options_new(
    enum aws_socket_type type,
    enum aws_socket_domain domain,
    uint32_t connect_timeout_ms,
    uint16_t keep_alive_interval_sec,
    uint16_t keep_alive_timeout_sec,
    uint16_t keep_alive_max_failed_probes,
    bool keepalive) {

    struct aws_allocator *allocator = aws_dotnet_get_allocator();
    struct aws_socket_options *options = aws_mem_calloc(allocator, 1, sizeof(struct aws_socket_options));
    if (!options) {
        aws_dotnet_throw_exception(aws_last_error(), "Failed to allocate new aws_socket_options");
        return NULL;
    }

    options->type = type;
    options->domain = domain;
    options->connect_timeout_ms = connect_timeout_ms;
    options->keep_alive_interval_sec = keep_alive_interval_sec;
    options->keep_alive_timeout_sec = keep_alive_timeout_sec;
    options->keep_alive_max_failed_probes = keep_alive_max_failed_probes;
    options->keepalive = keepalive;

    return options;
}

AWS_DOTNET_API
void aws_dotnet_socket_options_destroy(struct aws_socket_options *options) {
    if (options == NULL) {
        return;
    }

    struct aws_allocator *allocator = aws_dotnet_get_allocator();
    aws_mem_release(allocator, options);
}

File: aws-crt/IO/SocketOptions.cs
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */
using System;
using System.Security;
using System.Runtime.InteropServices;

namespace Aws.Crt.IO
{
    public enum SocketDomain {
        IPv4 = 0,
        IPv6 = 1,
        Local = 2
    }

    public enum SocketType {
        Stream = 0,
        Dgram = 1
    }

    public class SocketOptions
    {
        [SecuritySafeCritical]
        internal static class API
        {
            public delegate Handle aws_dotnet_socket_options_new(
                                    Int32 type, 
                                    Int32 domain, 
                                    UInt32 connect_timeout_ms, 
                                    UInt16 keep_alive_interval_sec, 
                                    UInt16 keep_alive_timeout_sec, 
                                    UInt16 keep_alive_max_failed_probes, 
                                    byte keepalive);
            public delegate void aws_dotnet_socket_options_destroy(IntPtr options);

            public static aws_dotnet_socket_options_new make_new = NativeAPI.Bind<aws_dotnet_socket_options_new>();
            public static aws_dotnet_socket_options_destroy destroy = NativeAPI.Bind<aws_dotnet_socket_options_destroy>();
        }

        internal class Handle : CRT.Handle
        {
            protected override bool ReleaseHandle()
            {
                API.destroy(handle);
                return true;
            }
        }

        internal Handle NativeHandle { 
            get {
                return API.make_new(
                            (Int32)Type,
                            (Int32)Domain,
                            ConnectTimeoutMs,
                            KeepAliveIntervalSeconds,
                            KeepAliveTimeoutSeconds,
                            0,
                            (byte)(KeepAlive ? 1: 0)
                );
            }        
        }

        public SocketType Type { get; set; }
        public SocketDomain Domain { get; set; }
        public UInt32 ConnectTimeoutMs { get; set; }
        public UInt16 KeepAliveIntervalSeconds { get; set; }
        public UInt16 KeepAliveTimeoutSeconds { get; set; }
        public bool KeepAlive { get; set; }
    }
}
